File: kotlin/bc-envelope/src/test/kotlin/com/blockchaincommons/bcenvelope/ObscuringTest.kt
package com.blockchaincommons.bcenvelope

import com.blockchaincommons.bccomponents.Digest
import com.blockchaincommons.bccomponents.SymmetricKey
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ObscuringTest {

    @Test
    fun testObscuring() {
        val key = SymmetricKey.create()
        val envelope = Envelope.from(PLAINTEXT_HELLO)
        assertFalse(envelope.isObscured())

        val encrypted = envelope.encryptSubject(key)
        assertTrue(encrypted.isObscured())

        val elided = envelope.elide()
        assertTrue(elided.isObscured())

        val compressed = envelope.compress()
        assertTrue(compressed.isObscured())

        // Cannot encrypt already encrypted
        assertTrue(
            try { encrypted.encryptSubject(key); false }
            catch (_: Exception) { true }
        )

        // Cannot encrypt elided
        assertTrue(
            try { elided.encryptSubject(key); false }
            catch (_: Exception) { true }
        )

        // OK to encrypt compressed
        val encryptedCompressed = compressed.encryptSubject(key)
        assertTrue(encryptedCompressed.isEncrypted())

        // OK to elide encrypted
        val elidedEncrypted = encrypted.elide()
        assertTrue(elidedEncrypted.isElided())

        // Eliding elided is idempotent
        val elidedElided = elided.elide()
        assertTrue(elidedElided.isElided())

        // OK to elide compressed
        val elidedCompressed = compressed.elide()
        assertTrue(elidedCompressed.isElided())

        // Cannot compress encrypted
        assertTrue(
            try { encrypted.compress(); false }
            catch (_: Exception) { true }
        )

        // Cannot compress elided
        assertTrue(
            try { elided.compress(); false }
            catch (_: Exception) { true }
        )

        // Compressing compressed is idempotent
        val compressedCompressed = compressed.compress()
        assertTrue(compressedCompressed.isCompressed())
    }

    @Test
    fun testNodesMatching() {
        val envelope = Envelope.from("Alice")
            .addAssertion("knows", "Bob")
            .addAssertion("age", 30)
            .addAssertion("city", "Boston")

        val knowsAssertion = envelope.assertionWithPredicate("knows")
        val knowsDigest = knowsAssertion.digest()

        val ageAssertion = envelope.assertionWithPredicate("age")
        val ageDigest = ageAssertion.digest()

        val elideTarget = setOf(knowsDigest)
        val compressTarget = setOf(ageDigest)

        var obscured = envelope.elideRemovingSet(elideTarget)
        obscured = obscured.elideRemovingSetWithAction(
            compressTarget,
            ObscureAction.Compress,
        )

        assertEquals(
            """
            "Alice" [
                "city": "Boston"
                COMPRESSED
                ELIDED
            ]
            """.trimIndent(),
            obscured.format()
        )

        // Test finding elided nodes
        val elidedNodes = obscured.nodesMatching(null, listOf(ObscureType.Elided))
        assertTrue(elidedNodes.contains(knowsDigest))

        // Test finding compressed nodes
        val compressedNodes = obscured.nodesMatching(null, listOf(ObscureType.Compressed))
        assertTrue(compressedNodes.contains(ageDigest))

        // Test with target filter
        val targetFilter = setOf(knowsDigest)
        val filtered = obscured.nodesMatching(targetFilter, listOf(ObscureType.Elided))
        assertEquals(1, filtered.size)
        assertTrue(filtered.contains(knowsDigest))
    }

    @Test
    fun testWalkUnelide() {
        val alice = Envelope.from("Alice")
        val bob = Envelope.from("Bob")
        val carol = Envelope.from("Carol")

        val envelope = Envelope.from("Alice")
            .addAssertion("knows", "Bob")
            .addAssertion("friend", "Carol")

        val elided = envelope
            .elideRemovingTarget(alice)
            .elideRemovingTarget(bob)

        assertEquals(
            """
            ELIDED [
                "friend": "Carol"
                "knows": ELIDED
            ]
            """.trimIndent(),
            elided.format()
        )

        val restored = elided.walkUnelide(listOf(alice, bob, carol))
        assertEquals(
            """
            "Alice" [
                "friend": "Carol"
                "knows": "Bob"
            ]
            """.trimIndent(),
            restored.format()
        )

        val partial = elided.walkUnelide(listOf(alice))
        assertEquals(
            """
            "Alice" [
                "friend": "Carol"
                "knows": ELIDED
            ]
            """.trimIndent(),
            partial.format()
        )

        val unchanged = elided.walkUnelide(emptyList())
        assertTrue(unchanged.isIdenticalTo(elided))
    }

    @Test
    fun testWalkDecrypt() {
        val key1 = SymmetricKey.create()
        val key2 = SymmetricKey.create()
        val key3 = SymmetricKey.create()

        val envelope = Envelope.from("Alice")
            .addAssertion("knows", "Bob")
            .addAssertion("age", 30)
            .addAssertion("city", "Boston")

        val knowsAssertion = envelope.assertionWithPredicate("knows")
        val ageAssertion = envelope.assertionWithPredicate("age")

        val encrypt1Target = setOf(knowsAssertion.digest())
        val encrypt2Target = setOf(ageAssertion.digest())

        val encrypted = envelope
            .elideRemovingSetWithAction(encrypt1Target, ObscureAction.Encrypt(key1))
            .elideRemovingSetWithAction(encrypt2Target, ObscureAction.Encrypt(key2))

        assertEquals(
            """
            "Alice" [
                "city": "Boston"
                ENCRYPTED (2)
            ]
            """.trimIndent(),
            encrypted.format()
        )

        // Decrypt with all keys
        val decrypted = encrypted.walkDecrypt(listOf(key1, key2))
        assertEquals(
            """
            "Alice" [
                "age": 30
                "city": "Boston"
                "knows": "Bob"
            ]
            """.trimIndent(),
            decrypted.format()
        )

        // Decrypt with wrong key unchanged
        val unchanged = encrypted.walkDecrypt(listOf(key3))
        assertTrue(unchanged.isIdenticalTo(encrypted))
    }

    @Test
    fun testWalkDecompress() {
        val envelope = Envelope.from("Alice")
            .addAssertion("knows", "Bob")
            .addAssertion("bio", "A".repeat(1000))
            .addAssertion("description", "B".repeat(1000))

        val bioAssertion = envelope.assertionWithPredicate("bio")
        val descAssertion = envelope.assertionWithPredicate("description")

        val bioDigest = bioAssertion.digest()
        val descDigest = descAssertion.digest()

        val compressTarget = setOf(bioDigest, descDigest)
        val compressed = envelope.elideRemovingSetWithAction(
            compressTarget,
            ObscureAction.Compress,
        )

        assertEquals(
            """
            "Alice" [
                "knows": "Bob"
                COMPRESSED (2)
            ]
            """.trimIndent(),
            compressed.format()
        )

        // Decompress all
        val decompressed = compressed.walkDecompress(null)
        assertTrue(decompressed.isEquivalentTo(envelope))

        // Decompress with target filter
        val target = setOf(bioDigest)
        val partial = compressed.walkDecompress(target)
        assertFalse(partial.isIdenticalTo(compressed))
        assertTrue(partial.isEquivalentTo(envelope))

        // Decompress with no match
        val noMatch = setOf(Digest.fromImage("nonexistent".toByteArray()))
        val unchangedResult = compressed.walkDecompress(noMatch)
        assertTrue(unchangedResult.isIdenticalTo(compressed))
    }

    @Test
    fun testMixedObscurationOperations() {
        val key = SymmetricKey.create()

        val envelope = Envelope.from("Alice")
            .addAssertion("knows", "Bob")
            .addAssertion("age", 30)
            .addAssertion("bio", "A".repeat(1000))

        val knowsAssertion = envelope.assertionWithPredicate("knows")
        val ageAssertion = envelope.assertionWithPredicate("age")
        val bioAssertion = envelope.assertionWithPredicate("bio")

        val knowsDigest = knowsAssertion.digest()
        val ageDigest = ageAssertion.digest()
        val bioDigest = bioAssertion.digest()

        val elideTarget = setOf(knowsDigest)
        val encryptTarget = setOf(ageDigest)
        val compressTarget = setOf(bioDigest)

        val obscured = envelope
            .elideRemovingSet(elideTarget)
            .elideRemovingSetWithAction(encryptTarget, ObscureAction.Encrypt(key))
            .elideRemovingSetWithAction(compressTarget, ObscureAction.Compress)

        val elidedNodes = obscured.nodesMatching(null, listOf(ObscureType.Elided))
        val encryptedNodes = obscured.nodesMatching(null, listOf(ObscureType.Encrypted))
        val compressedNodes = obscured.nodesMatching(null, listOf(ObscureType.Compressed))

        assertTrue(elidedNodes.contains(knowsDigest))
        assertTrue(encryptedNodes.contains(ageDigest))
        assertTrue(compressedNodes.contains(bioDigest))

        // Restore everything
        val restored = obscured
            .walkUnelide(listOf(knowsAssertion))
            .walkDecrypt(listOf(key))
            .walkDecompress(null)
        assertTrue(restored.isEquivalentTo(envelope))
    }
}

File: go/bcenvelope/obscuring_test.go
package bcenvelope

import (
	"strings"
	"testing"

	bccomponents "github.com/nickel-blockchaincommons/bccomponents-go"
)

func TestObscuring(t *testing.T) {
	key := bccomponents.NewSymmetricKey()

	envelope := NewEnvelope(plaintextHello)
	if envelope.IsObscured() {
		t.Fatal("plain envelope should not be obscured")
	}

	// Encrypted
	encrypted, err := envelope.EncryptSubject(key)
	if err != nil {
		t.Fatalf("encrypt failed: %v", err)
	}
	if !encrypted.IsObscured() {
		t.Fatal("encrypted should be obscured")
	}

	// Elided
	elided := envelope.Elide()
	if !elided.IsObscured() {
		t.Fatal("elided should be obscured")
	}

	// Compressed
	compressed, err := envelope.Compress()
	if err != nil {
		t.Fatalf("compress failed: %v", err)
	}
	if !compressed.IsObscured() {
		t.Fatal("compressed should be obscured")
	}

	// ENCRYPTION

	// Cannot encrypt an encrypted envelope.
	_, err = encrypted.EncryptSubject(key)
	if err == nil {
		t.Fatal("expected error encrypting encrypted envelope")
	}

	// Cannot encrypt an elided envelope.
	_, err = elided.EncryptSubject(key)
	if err == nil {
		t.Fatal("expected error encrypting elided envelope")
	}

	// OK to encrypt a compressed envelope.
	encryptedCompressed, err := compressed.EncryptSubject(key)
	if err != nil {
		t.Fatalf("encrypt compressed failed: %v", err)
	}
	if !encryptedCompressed.IsEncrypted() {
		t.Fatal("encrypted compressed should be encrypted")
	}

	// ELISION

	// OK to elide an encrypted envelope.
	elidedEncrypted := encrypted.Elide()
	if !elidedEncrypted.IsElided() {
		t.Fatal("elided encrypted should be elided")
	}

	// Eliding an elided envelope is idempotent.
	elidedElided := elided.Elide()
	if !elidedElided.IsElided() {
		t.Fatal("elided elided should be elided")
	}

	// OK to elide a compressed envelope.
	elidedCompressed := compressed.Elide()
	if !elidedCompressed.IsElided() {
		t.Fatal("elided compressed should be elided")
	}

	// COMPRESSION

	// Cannot compress an encrypted envelope.
	_, err = encrypted.Compress()
	if err == nil {
		t.Fatal("expected error compressing encrypted envelope")
	}

	// Cannot compress an elided envelope.
	_, err = elided.Compress()
	if err == nil {
		t.Fatal("expected error compressing elided envelope")
	}

	// Compressing a compressed envelope is idempotent.
	compressedCompressed, err := compressed.Compress()
	if err != nil {
		t.Fatalf("compress compressed failed: %v", err)
	}
	if !compressedCompressed.IsCompressed() {
		t.Fatal("compressed compressed should be compressed")
	}
}

func TestNodesMatching(t *testing.T) {
	envelope := NewEnvelope("Alice").
		AddAssertion("knows", "Bob").
		AddAssertion("age", 30).
		AddAssertion("city", "Boston")

	// Get some digests for targeting
	knowsAssertion, err := envelope.AssertionWithPredicate("knows")
	if err != nil {
		t.Fatalf("knows assertion not found: %v", err)
	}
	knowsDigest := knowsAssertion.Digest()

	ageAssertion, err := envelope.AssertionWithPredicate("age")
	if err != nil {
		t.Fatalf("age assertion not found: %v", err)
	}
	ageDigest := ageAssertion.Digest()

	// Elide one assertion, compress another
	elideTarget := make(map[bccomponents.Digest]struct{})
	elideTarget[knowsDigest] = struct{}{}

	compressTarget := make(map[bccomponents.Digest]struct{})
	compressTarget[ageDigest] = struct{}{}

	obscured := envelope.ElideRemovingSet(elideTarget)
	obscured = obscured.ElideRemovingSetWithAction(
		compressTarget,
		ObscureActionCompress,
		nil,
	)

	assertActualExpected(t, obscured.Format(), `"Alice" [
    "city": "Boston"
    COMPRESSED
    ELIDED
]`)

	// Test finding elided nodes
	elidedNodes := obscured.NodesMatching(nil, []ObscureType{ObscureTypeElided})
	if _, ok := elidedNodes[knowsDigest]; !ok {
		t.Fatal("expected knows digest in elided nodes")
	}

	// Test finding compressed nodes
	compressedNodes := obscured.NodesMatching(nil, []ObscureType{ObscureTypeCompressed})
	if _, ok := compressedNodes[ageDigest]; !ok {
		t.Fatal("expected age digest in compressed nodes")
	}

	// Test finding with target filter
	targetFilter := make(map[bccomponents.Digest]struct{})
	targetFilter[knowsDigest] = struct{}{}
	filtered := obscured.NodesMatching(targetFilter, []ObscureType{ObscureTypeElided})
	if len(filtered) != 1 {
		t.Fatalf("expected 1 filtered, got %d", len(filtered))
	}
	if _, ok := filtered[knowsDigest]; !ok {
		t.Fatal("expected knows digest in filtered")
	}

	// Test finding all obscured nodes (no type filter)
	allInTarget := obscured.NodesMatching(elideTarget, nil)
	if len(allInTarget) != 1 {
		t.Fatalf("expected 1 in target, got %d", len(allInTarget))
	}
	if _, ok := allInTarget[knowsDigest]; !ok {
		t.Fatal("expected knows digest in allInTarget")
	}

	// Test with no matches
	noMatchTarget := make(map[bccomponents.Digest]struct{})
	noMatchDigest := bccomponents.DigestFromImage([]byte("nonexistent"))
	noMatchTarget[noMatchDigest] = struct{}{}
	noMatches := obscured.NodesMatching(noMatchTarget, []ObscureType{ObscureTypeElided})
	if len(noMatches) != 0 {
		t.Fatalf("expected 0 no-matches, got %d", len(noMatches))
	}
}

func TestWalkUnelide(t *testing.T) {
	alice := NewEnvelope("Alice")
	bob := NewEnvelope("Bob")
	carol := NewEnvelope("Carol")

	envelope := NewEnvelope("Alice").
		AddAssertion("knows", "Bob").
		AddAssertion("friend", "Carol")

	// Elide multiple parts
	elided := envelope.
		ElideRemovingTarget(alice).
		ElideRemovingTarget(bob)

	// Verify parts are elided
	assertActualExpected(t, elided.Format(), `ELIDED [
    "friend": "Carol"
    "knows": ELIDED
]`)

	// Restore with walk_unelide
	restored := elided.WalkUnelide([]*Envelope{alice, bob, carol})

	assertActualExpected(t, restored.Format(), `"Alice" [
    "friend": "Carol"
    "knows": "Bob"
]`)

	// Test with partial restoration (only some envelopes provided)
	partial := elided.WalkUnelide([]*Envelope{alice})
	assertActualExpected(t, partial.Format(), `"Alice" [
    "friend": "Carol"
    "knows": ELIDED
]`)

	// Test with no matching envelopes
	unchanged := elided.WalkUnelide(nil)
	if !unchanged.IsIdenticalTo(elided) {
		t.Fatal("unchanged should be identical to elided")
	}
}

func TestWalkDecrypt(t *testing.T) {
	key1 := bccomponents.NewSymmetricKey()
	key2 := bccomponents.NewSymmetricKey()
	key3 := bccomponents.NewSymmetricKey()

	envelope := NewEnvelope("Alice").
		AddAssertion("knows", "Bob").
		AddAssertion("age", 30).
		AddAssertion("city", "Boston")

	// Encrypt different parts with different keys
	knowsAssertion, err := envelope.AssertionWithPredicate("knows")
	if err != nil {
		t.Fatalf("knows assertion not found: %v", err)
	}
	ageAssertion, err := envelope.AssertionWithPredicate("age")
	if err != nil {
		t.Fatalf("age assertion not found: %v", err)
	}

	encrypt1Target := make(map[bccomponents.Digest]struct{})
	encrypt1Target[knowsAssertion.Digest()] = struct{}{}

	encrypt2Target := make(map[bccomponents.Digest]struct{})
	encrypt2Target[ageAssertion.Digest()] = struct{}{}

	encrypted := envelope.ElideSetWithAction(
		encrypt1Target, false, ObscureActionEncrypt, &ObscureActionEncryptWithKey{Key: key1},
	)
	encrypted = encrypted.ElideSetWithAction(
		encrypt2Target, false, ObscureActionEncrypt, &ObscureActionEncryptWithKey{Key: key2},
	)

	// Verify parts are encrypted
	assertActualExpected(t, encrypted.Format(), `"Alice" [
    "city": "Boston"
    ENCRYPTED (2)
]`)

	// Decrypt with all keys
	decrypted := encrypted.WalkDecrypt([]bccomponents.SymmetricKey{key1, key2})
	assertActualExpected(t, decrypted.Format(), `"Alice" [
    "age": 30
    "city": "Boston"
    "knows": "Bob"
]`)

	// Decrypt with only one key (partial decryption)
	partial := encrypted.WalkDecrypt([]bccomponents.SymmetricKey{key1})
	if partial.IsIdenticalTo(encrypted) {
		t.Fatal("partial should differ from encrypted")
	}
	// Note: partial is still equivalent because encrypted nodes preserve digests
	if !partial.IsEquivalentTo(envelope) {
		t.Fatal("partial should be equivalent to original")
	}

	// There should still be one encrypted node remaining
	partialFormat := partial.Format()
	if !strings.Contains(partialFormat, "ENCRYPTED") {
		t.Fatal("expected ENCRYPTED in partial format")
	}
	if !strings.Contains(partialFormat, `"knows": "Bob"`) {
		t.Fatal("expected 'knows': 'Bob' in partial format")
	}

	// Decrypt with wrong key (should be unchanged)
	unchanged := encrypted.WalkDecrypt([]bccomponents.SymmetricKey{key3})
	if !unchanged.IsIdenticalTo(encrypted) {
		t.Fatal("unchanged should be identical to encrypted")
	}
}

func TestWalkDecompress(t *testing.T) {
	envelope := NewEnvelope("Alice").
		AddAssertion("knows", "Bob").
		AddAssertion("bio", strings.Repeat("A", 1000)).
		AddAssertion("description", strings.Repeat("B", 1000))

	// Compress multiple parts
	bioAssertion, err := envelope.AssertionWithPredicate("bio")
	if err != nil {
		t.Fatalf("bio assertion not found: %v", err)
	}
	descAssertion, err := envelope.AssertionWithPredicate("description")
	if err != nil {
		t.Fatalf("description assertion not found: %v", err)
	}

	bioDigest := bioAssertion.Digest()
	descDigest := descAssertion.Digest()

	compressTarget := make(map[bccomponents.Digest]struct{})
	compressTarget[bioDigest] = struct{}{}
	compressTarget[descDigest] = struct{}{}

	compressed := envelope.ElideRemovingSetWithAction(
		compressTarget,
		ObscureActionCompress,
		nil,
	)

	// Verify parts are compressed
	assertActualExpected(t, compressed.Format(), `"Alice" [
    "knows": "Bob"
    COMPRESSED (2)
]`)

	// Decompress all
	decompressed := compressed.WalkDecompress(nil)
	if !decompressed.IsEquivalentTo(envelope) {
		t.Fatal("decompressed should be equivalent to original")
	}

	// Decompress with target filter (only one node)
	target := make(map[bccomponents.Digest]struct{})
	target[bioDigest] = struct{}{}

	partialDecomp := compressed.WalkDecompress(target)
	if partialDecomp.IsIdenticalTo(compressed) {
		t.Fatal("partial should differ from compressed")
	}
	// Note: partial is still equivalent because compressed nodes preserve digests
	if !partialDecomp.IsEquivalentTo(envelope) {
		t.Fatal("partial should be equivalent to original")
	}

	// Bio should be decompressed but description still compressed
	stillCompressed := partialDecomp.NodesMatching(nil, []ObscureType{ObscureTypeCompressed})
	if _, ok := stillCompressed[descDigest]; !ok {
		t.Fatal("expected description still compressed")
	}
	if _, ok := stillCompressed[bioDigest]; ok {
		t.Fatal("expected bio to be decompressed")
	}

	// Decompress with non-matching target (should be unchanged)
	noMatch := make(map[bccomponents.Digest]struct{})
	noMatchDigest := bccomponents.DigestFromImage([]byte("nonexistent"))
	noMatch[noMatchDigest] = struct{}{}
	unchanged := compressed.WalkDecompress(noMatch)
	if !unchanged.IsIdenticalTo(compressed) {
		t.Fatal("unchanged should be identical to compressed")
	}
}

func TestMixedObscurationOperations(t *testing.T) {
	key := bccomponents.NewSymmetricKey()

	envelope := NewEnvelope("Alice").
		AddAssertion("knows", "Bob").
		AddAssertion("age", 30).
		AddAssertion("bio", strings.Repeat("A", 1000))

	knowsAssertion, err := envelope.AssertionWithPredicate("knows")
	if err != nil {
		t.Fatalf("knows assertion not found: %v", err)
	}
	ageAssertion, err := envelope.AssertionWithPredicate("age")
	if err != nil {
		t.Fatalf("age assertion not found: %v", err)
	}
	bioAssertion, err := envelope.AssertionWithPredicate("bio")
	if err != nil {
		t.Fatalf("bio assertion not found: %v", err)
	}

	knowsDigest := knowsAssertion.Digest()
	ageDigest := ageAssertion.Digest()
	bioDigest := bioAssertion.Digest()

	// Apply different obscuration types
	elideTarget := make(map[bccomponents.Digest]struct{})
	elideTarget[knowsDigest] = struct{}{}

	encryptTarget := make(map[bccomponents.Digest]struct{})
	encryptTarget[ageDigest] = struct{}{}

	compressTarget := make(map[bccomponents.Digest]struct{})
	compressTarget[bioDigest] = struct{}{}

	obscured := envelope.ElideRemovingSet(elideTarget)
	obscured = obscured.ElideRemovingSetWithAction(
		encryptTarget,
		ObscureActionEncrypt,
		&ObscureActionEncryptWithKey{Key: key},
	)
	obscured = obscured.ElideRemovingSetWithAction(
		compressTarget,
		ObscureActionCompress,
		nil,
	)

	// Verify different obscuration types
	elidedNodes := obscured.NodesMatching(nil, []ObscureType{ObscureTypeElided})
	encryptedNodes := obscured.NodesMatching(nil, []ObscureType{ObscureTypeEncrypted})
	compressedNodes := obscured.NodesMatching(nil, []ObscureType{ObscureTypeCompressed})

	if _, ok := elidedNodes[knowsDigest]; !ok {
		t.Fatal("expected knows in elided nodes")
	}
	if _, ok := encryptedNodes[ageDigest]; !ok {
		t.Fatal("expected age in encrypted nodes")
	}
	if _, ok := compressedNodes[bioDigest]; !ok {
		t.Fatal("expected bio in compressed nodes")
	}

	// Restore everything
	restored := obscured.
		WalkUnelide([]*Envelope{knowsAssertion}).
		WalkDecrypt([]bccomponents.SymmetricKey{key}).
		WalkDecompress(nil)

	if !restored.IsEquivalentTo(envelope) {
		t.Fatal("restored should be equivalent to original")
	}
}
